hpat/distributed_analysis.py

```diff
@@ -298,10 +299,51 @@ def _analyze_call(self, lhs, rhs, func_var, args, array_dists):
                 self._meet_array_dists(lhs, args[0].name, array_dists)
                 return
 
+        if isinstance(func_mod, ir.Var) and self._analyze_call_d4p(lhs, func_name, self.typemap[func_mod.name], args, array_dists):
+            return
+
         # set REP if not found
         self._analyze_call_set_REP(lhs, args, array_dists)
 
 
+    def _analyze_call_d4p(self, lhs, func_name, mod_name, args, array_dists):
+        '''
+        Analyze distribution for calls to daal4py.
+        Return True of a call for daal4py was detected and handled.
+        We cannot simply "meet" distributions, the d4p algos accept a certain decomposition only.
+        The required distribution/decomposition is defined in the algorithms specs.
+        We raise an exception if the required distribution cannot be met.
+        '''
+        if func_name == 'compute':
+            # every d4p algo gets executed by invoking "compute".
+            # we need to find the algorithm that's currently called
+            for algo in d4p_algos:
+                if algo.all_nbtypes[algo.name] == mod_name:
+                    # handle all input arguments and set their distribution as given by the spec
+                    for i in range(len(args)):
+                        aname = args[i].name
+                        adist = algo.spec.input_types[i][2]
+                        if aname not in array_dists:
+                            array_dists[aname] = adist
+                        else:
+                            # bail out if there is a distribution conflict with some other use of the argument
+                            # FIXME: handle Distribution.Thread and Disribution.REP as equivalent
+                            assert array_dists[aname] == adist,\
```

Comment:
I think a better way than this assertion here is probably "(meet of array_dists[aname] and adist) == adist". This enables converting argument distribution to a lower one without problem. For example, if an argument is 1D but adist is REP, the argument can be converted to REP.

Reply:
Not sure what this would imply. Right now, it assumes the data is OneD and act accordingly. 
I don't know which of the algorithms are sensitive to data duplication and which are not.
It requires explicit configuration to switch to proper REP processing in a distributed setup. Doable, but should probably be postponed.